### ISSUE TITLE
Schannel: Data size exceeded in callback

### ISSUE THREAD

**Issue by gvanem** [curl/curl#6462]
### I did this
Built libcurl + curl with `-DDEBUGBUILD` and issued these commands:

```
c:\> set CURL_SSL_BACKEND=schannel
c:\> curl.exe https://www.vg.no
```
and curl failed with:
```
...
* schannel: failed to decrypt data, need more data
* schannel: schannel_recv cleanup
* schannel: decrypted data returned 32633
* schannel: decrypted data buffer: offset 0 length 32768
Warning: Data size (32633) exceeds single call write limit of 16384!
...

curl: (23) Failed writing received data to disk/application
```

### I expected the following
To receive the whole 400 kByte page.

### curl/libcurl version
```
c:\> curl -V
curl 7.75.0-DEV (i386-pc-win32) libcurl/7.75.0-DEV (wolfSSL/4.5.0) (OpenSSL/3.0.0) Schannel zlib/1.2.11 brotli/1.0.0 zstd/1.4.4 c-ares/1.17.0 WinIDN libssh2/1.9.0_DEV librtmp/2.3 Hyper/0.13.8 
Release-Date: [unreleased]
Protocols: dict file ftp ftps gopher gophers http https imap imaps mqtt pop3 pop3s rtmp scp sftp smb smbs smtp smtps telnet tftp
Features: AsynchDNS Debug HTTP2 IDN IPv6 Kerberos Largefile MultiSSL NTLM SPNEGO SSL SSPI TLS-SRP TrackMemory UnixSockets alt-svc brotli libz zstd
```

### operating system

Windows-10, MSVC-2019 or clang-cl v11.

The above Warning is inside a DEBUGBUILD block [here](https://github.com/curl/curl/blob/master/src/tool_cb_wrt.c#L146). I patched `tool_cb_wrt.c` to show the sizes shown above.

Building w/o `-DDEBUGBUILD`, all seems okay.

Also, building with:
```
 CFLAGS += -DDEBUGBUILD                 \
           -DCURL_MAX_READ_SIZE=2000000 \
           -DCURL_MAX_WRITE_SIZE=2000000
```
is also fine.

Linking libcurl.dll with my Wsock-trace library shows the last `recv()` prior to the problem was
returning 33532 bytes:
```
* schannel: encdata_buffer resized 33792
* schannel: encrypted data buffer: offset 260 length 33792
  * 2.743 sec: f:/MingW32/src/inet/curl/lib/memdebug.c(328) (curl_dbg_recv+138):
    recv (796, 0x0DC2EB7C, 33532, 0) --> 33532 bytes.
```
Too much for `CURL_MAX_READ_SIZE` ?

I also noted a `* schannel: client wants to read 32768 bytes`. Where is that called from?

**Comment by jay** [comment]
I can't reproduce this.

master 13fe0b6 2020-01-11
~~~
curl 7.75.0-DEV (i386-pc-win32) libcurl/7.75.0-DEV (OpenSSL/1.1.1i) Schannel
Release-Date: [unreleased]
Protocols: dict file ftp ftps gopher gophers http https imap imaps ldap ldaps mqtt pop3 pop3s rtsp smb smbs smtp smtps telnet tftp
Features: AsynchDNS Debug IPv6 Kerberos Largefile MultiSSL NTLM SPNEGO SSL SSPI Unicode UnixSockets alt-svc
~~~

> Warning: Data size (32633) exceeds single call write limit of 16384!

I don't see how that could happen unless CURL_MAX_WRITE_SIZE is defined differently in libcurl than it is in the curl tool. libcurl sends chunks that do not exceed CURL_MAX_WRITE_SIZE to the write callback.

https://github.com/curl/curl/blob/e052859759b34d0e05ce0f17244873e5cd7b457b/lib/sendf.c#L534-L542

https://github.com/curl/curl/blob/e052859759b34d0e05ce0f17244873e5cd7b457b/src/tool_cb_wrt.c#L145-L148

> I also noted a `* schannel: client wants to read 32768 bytes`. Where is that called from?

That's fine. It's from [schannel_recv](https://github.com/curl/curl/blob/curl-7_74_0/lib/vtls/schannel.c#L1769).

**Comment by gvanem** [comment]
>  libcurl sends chunks that do not exceed CURL_MAX_WRITE_SIZE to the write callback.

Yes, I assumed that. But adding a trace inside the ` Chop data, write chunks` loop, I see it's *not* called.

Forgot to say earlier I built with `-DUSE_HYPER`. Now when removing that I have no problem with Schannel and `-DCURLDEBUG`.

So building with  `-DUSE_HYPER -DCURLDEBUG` and adding a similar trace to `c-hyper.c`, I see 
`hyper_body_chunk()` is to blame for this oversized chunk:
```
* schannel: decrypted data buffer: offset 0 length 32768
{ [22392 bytes data]
hyper_body_chunk(): len:22392
Warning: Data size (22392) exceeds single call write limit of 16384!
```

The length from `hyper_buf_len()` probably?

**Comment by gvanem** [comment]
Looking into c-hyper I failed to see how it should cap the length of a `hyper_buf`. I seems to be the responsibility of the user,   i.e. libcurl's no? So with this patch:
```diff
--- a/c-hyper.c 2021-01-14 15:27:43
+++ b/c-hyper.c 2021-01-17 10:36:23
@@ -67,6 +67,8 @@

   (void)ctx;

+  buflen = min (buflen, CURL_MAX_WRITE_SIZE);
+
   result = Curl_read(conn, conn->sockfd, (char *)buf, buflen, &nread);
```

I see ` tool_write_cb()` is happy and my issue is gone. Does this patch makes sense?

**Comment by bagder** [comment]
Shouldn't the fix rather be to make sure the write callback in `hyper_body_chunk` doesn't send more data than documented?

**Comment by gvanem** [comment]
> .... hyper_body_chunk doesn't send more data than documented?

*documented*? Where? The subject should be about Hyper. But since it's experimental, it's hard to tell how it can create bug all over the place. Besides, the code-style, makes it really hard to follow the flow of things.

**Comment by bagder** [comment]
The [man page for CURLOPT_WRITEFUNCTION](https://curl.se/libcurl/c/CURLOPT_WRITEFUNCTION.html) says

*"The maximum amount of body data that will be passed to the write callback is defined in the curl.h header file: CURL_MAX_WRITE_SIZE"*

> it's hard to tell how it can create bug all over the place

I don't understand. It's quite a large change, I don't think it's strange it comes with a fair amount of wrinkles to iron out.

> Besides, the code-style, makes it really hard to follow the flow of things.

Code-style how? How would you suggest we'd improve it?

**Comment by gvanem** [comment]
> "The maximum amount of body data that will be passed to the write callback is defined in the curl.h header file: 
>  CURL_MAX_WRITE_SIZE"

I knew that already since I wrote,
> Too much for CURL_MAX_READ_SIZE ?

but confused `CURL_MAX_WRITE_SIZE` with `CURL_MAX_READ_SIZE`.

> Code-style how? How would you suggest we'd improve it?

One thing are things like these:
```
      wrote = writebody(ptr, 1, chunklen, data->set.out);
```
Why not make it plain that it's a call via a function-pointer:
```
      wrote = (*writebody)(ptr, 1, chunklen, data->set.out);
```

**Comment by bagder** [comment]
> Why not make it plain that it's a call via a function-pointer:

Just a matter of taste. And #6473 now removes that from c-hyper.c...
